src/main/java/org/opencds/cqf/providers/FHIREndpointProvider.java

```diff
@@ -0,0 +1,149 @@
+import ca.uhn.fhir.rest.client.api.IGenericClient;
+import ca.uhn.fhir.rest.client.api.ServerValidationModeEnum;
+import ca.uhn.fhir.rest.client.interceptor.BasicAuthInterceptor;
+import ca.uhn.fhir.rest.param.StringAndListParam;
+import ca.uhn.fhir.rest.param.StringOrListParam;
+import ca.uhn.fhir.rest.param.StringParam;
+import org.hl7.fhir.dstu3.model.*;
+import org.opencds.cqf.exceptions.Helper;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+
+public class FHIREndpointProvider extends EndpointResourceProvider {
+
+    private JpaDataProvider provider;
+    private IFhirSystemDao systemDao;
+
+    public FHIREndpointProvider(JpaDataProvider provider, IFhirSystemDao systemDao) {
+        this.provider = provider;
+        this.systemDao = systemDao;
+    }
+
+    @Operation(name="cache-valuesets", idempotent = true)
+    public Resource cacheValuesets(
+            RequestDetails details,
+            @IdParam IdType theId,
+            @RequiredParam(name="valuesets") StringAndListParam valuesets,
+            @OptionalParam(name="user") String userName,
+            @OptionalParam(name="pass") String password
+    ) {
+
+        Endpoint endpoint = this.getDao().read(theId);
+
+        if (endpoint == null) {
+            return Helper.createErrorOutcome("Could not find Endpoint/" + theId);
+        }
+
+        provider.setEndpoint(endpoint.getAddress());
+        provider.getFhirContext().getRestfulClientFactory().setServerValidationMode(ServerValidationModeEnum.NEVER);
+        IGenericClient client = provider.getFhirClient();
+
+        if (userName != null || password != null) {
+            if (userName == null) {
+                Helper.createErrorOutcome("Password was provided, but not a user name.");
+            }
+            else if (password == null) {
+                Helper.createErrorOutcome("User name was provided, but not a password.");
+            }
+
+            BasicAuthInterceptor basicAuth = new BasicAuthInterceptor(userName, password);
+            client.registerInterceptor(basicAuth);
+
+            // TODO - more advanced security like bearer tokens, etc...
+        }
+
+        try {
+            Bundle bundleToPost = new Bundle();
+            for (StringOrListParam params : valuesets.getValuesAsQueryTokens()) {
+                for (StringParam valuesetId : params.getValuesAsQueryTokens()) {
+                    bundleToPost.addEntry()
+                            .setRequest(new Bundle.BundleEntryRequestComponent().setMethod(Bundle.HTTPVerb.PUT).setUrl("ValueSet/" + valuesetId.getValue()))
+                            .setResource(resolveValueSet(client, valuesetId.getValue()));
+                }
+            }
+
+            return (Resource) systemDao.transaction(details, bundleToPost);
+        } catch (Exception e) {
+            return Helper.createErrorOutcome(e.getMessage());
+        }
+    }
+
+    private ValueSet cleanExpandedValueSet(ValueSet expandedValueSet) {
```

Comment:
Should this be named "getCachedValueSet"? That's what this is doing, creating the "cached" version of the ValueSet.

Reply:
Yeah, I'll rename that.